File: trie.cpp
#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

struct Node {
    int idx[26];
    bool finish;
};

struct Trie {
    vector<Node> nodes{Node()};

    void insert(const string s) {
        int cur = 0;

        for (char c : s) {
            if (!nodes[cur].idx[c - 97]) {
                nodes[cur].idx[c - 97] = nodes.size();
                nodes.push_back(Node());
            }

            cur = nodes[cur].idx[c - 97];
        }

        nodes[cur].finish = true;
    }

    bool search(const string s) {
        int cur = 0;

        for (char c : s) {
            if (!nodes[cur].idx[c - 97])
                return false;

            cur = nodes[cur].idx[c - 97];
        }

        return nodes[cur].finish;
    }
};

//--------------------------------------------------------------------------------

struct Trie {
    map<char, Trie *> nxt;
    bool finish;

    Trie() { finish = false; }

    ~Trie() { nxt.clear(); }

    bool insert(const char *s) {
        if (*s == '\0') {
            finish = true;
            return !nxt.sz;
        }

        if (finish)
            return false;

        if (nxt.find(*s) == nxt.end())
            nxt[*s] = new Trie;

        return nxt[*s]->insert(s + 1);
    }
};

//--------------------------------------------------------------------------------

struct Trie {
    vector<pair<char, Trie *>> nxt;
    bool finish;

    Trie() { finish = false; }

    ~Trie() { nxt.clear(); }

    void insert(const char *s) {
        if (*s == '\0') {
            finish = true;
            return;
        }

        for (auto node : nxt) {
            if (node.fi == *s) {
                node.se->insert(s + 1);
                return;
            }
        }

        nxt.push_back({*s, new Trie});
        nxt.bk.se->insert(s + 1);
    }
};

File: python/trie.py
class Node:
    def __init__(self, key, data=None):
        self.key = key
        self.data = data
        self.child = {}


class Trie:
    def __init__(self):
        self.head = Node(None)

    def insert(self, string):
        cur_node = self.head
        for char in string:
            if char not in cur_node.child:
                cur_node.child[char] = Node(char)
            cur_node = cur_node.child[char]
        cur_node.data = string

    def search(self, string):
        cur_node = self.head
        for char in string:
            if char not in cur_node.child:
                return False
            cur_node = cur_node.child[char]
        if cur_node.data is not None:
            return True
        return False

#--------------------------------------------------------------------------------

class Node:
    def __init__(self):
        self.idx = [0] * 26


class Trie:
    def __init__(self):
        self.nodes = [Node()]

    def insert(self, string):
        cur = 0
        for char in string:
            char = ord(char) - 97
            if not self.nodes[cur].idx[char]:
                self.nodes[cur].idx[char] = len(self.nodes)
                self.nodes.append(Node())
            
            cur = self.nodes[cur].idx[char]
